src/transport/PointToPointBroker.cpp

```diff
@@ -564,10 +643,54 @@ std::vector<uint8_t> PointToPointBroker::recvMessage(int groupId,
                      recvEndpoints[label]->getAddress());
     }
 
-    Message message = recvEndpoints[label]->recv();
+    return recvEndpoints[label]->recv();
+}
 
-    // TODO - possible to avoid this copy?
-    return message.dataCopy();
+std::vector<uint8_t> PointToPointBroker::recvMessage(int groupId,
+                                                     int sendIdx,
+                                                     int recvIdx)
+{
+    // If we don't need to receive messages in order, return here
+    if (!orderMsg.load(std::memory_order_acquire)) {
+        // TODO - can we avoid this copy?
+        return doRecvMessage(groupId, sendIdx, recvIdx).dataCopy();
+    }
+
+    // Get the sequence number we expect to receive
+    int expectedSeqNum = getExpectedSeqNum(groupId, sendIdx);
+
+    // We first check if we have already received the message. We only need to
+    // check this once.
+    auto wantedIt =
+      std::find_if(outOfOrderMsgs.begin(),
+                   outOfOrderMsgs.end(),
+                   [expectedSeqNum](const Message& msg) {
+                       return msg.getSequenceNum() == expectedSeqNum;
+                   });
+    if (wantedIt != outOfOrderMsgs.end()) {
+        incrementRecvMsgCount(groupId, sendIdx);
+        Message returnMsg = std::move(*wantedIt);
+        outOfOrderMsgs.erase(wantedIt);
+        return returnMsg.dataCopy();
+    }
+
+    // Given that we don't have the message, we query the transport layer until
+    // we receive it
+    while (true) {
```

Comment:
`while(true)` loops hanging are quite difficult to debug if there's no logging. I think we should have some trace logging on entering the while loop and receiving out-of-order messages.